File: python/taflow/metrics/profit_factor.py
"""Gross-profit to gross-loss quality metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import ProfitFactor as _Native
from ._input import as_metric_series


class ProfitFactor:
    """Compute gross positive observations divided by absolute gross losses.

    Strictly positive values contribute to gross profit, strictly negative
    values contribute to signed gross loss, and exact zeros contribute to
    neither sum. The result matches QuantStats 0.0.81 ``profit_factor`` with
    ``prepare_returns=False`` for defined finite cases; this is the independent
    executable oracle mapping. TAFlow deliberately
    normalizes empty and all-zero input to ``None``. Positive-only input returns
    positive infinity, while loss-only input returns ``0.0``.

    ``from_returns`` consumes decimal simple period returns. ``from_pnl``
    consumes non-cumulative raw period P&L and accepts no initial capital;
    ``from_trades`` consumes realized P&L for closed trades. Values supplied to
    P&L and trade input methods are never converted or annualized. NaNs are omitted
    by default or rejected by ``nan_policy="raise"``; infinities are always
    rejected. Warm-up requires one nonzero usable observation. Mutating lifecycle
    methods are fluent, native bulk work releases the GIL, and Rust owns all
    allocation-free O(1) arithmetic. vectorbt is a documented trade-level
    cross-check but is not installed in the pinned test environment.
    """

    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_returns(
        self, returns: Any, *, column: str | None = None
    ) -> "ProfitFactor":
        """Append chronological returns observations and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def from_pnl(
        self, pnl: Any, *, column: str | None = None
    ) -> "ProfitFactor":
        """Append chronological pnl observations and return this metric."""
        self._state.from_pnl(as_metric_series(pnl, column=column))
        return self

    def from_trades(
        self, trades: Any, *, column: str | None = None
    ) -> "ProfitFactor":
        """Append chronological trades observations and return this metric."""
        self._state.from_trades(as_metric_series(trades, column=column))
        return self

    def append(self, value: float) -> "ProfitFactor":
        """Append one value in the selected domain and return this metric."""
        self._state.append(float(value))
        return self

    def extend(
        self, values: Any, *, column: str | None = None
    ) -> "ProfitFactor":
        """Append chronological observations and return this metric."""
        self._state.extend(as_metric_series(values, column=column))
        return self

    @property
    def value(self) -> float | None:
        """Return profit factor, including positive infinity when unbounded."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current scalar without replaying processed input."""
        return self._state.compute()

    def reset(self) -> "ProfitFactor":
        """Clear observations, preserve input configuration, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable observations processed by Rust."""
        return len(self._state)


__all__ = ["ProfitFactor"]

File: crates/taflow-metrics/src/metrics/profit_factor.rs
use crate::{
    primitives::GainLossState, MetricError, MetricInputKind, MetricInputState, MetricResult,
    NanPolicy,
};

/// Gross positive observations divided by absolute gross negative observations.
#[derive(Debug, Clone)]
pub struct ProfitFactor {
    input: MetricInputState,
    observations: GainLossState,
}

impl ProfitFactor {
    /// Construct an empty state for returns, raw period P&L, or closed trades.
    pub fn new(input_kind: MetricInputKind, nan_policy: NanPolicy) -> MetricResult<Self> {
        if !matches!(
            input_kind,
            MetricInputKind::Returns | MetricInputKind::RawPnl | MetricInputKind::Trades
        ) {
            return Err(MetricError::InvalidParameter {
                name: "input_kind",
                value: format!("{input_kind:?}"),
                reason:
                    "profit factor requires simple returns, raw period P&L, or closed-trade P&L",
            });
        }
        Ok(Self {
            input: MetricInputState::new(input_kind, nan_policy)?,
            observations: GainLossState::new(),
        })
    }

    /// Append one observation and return the current profit factor.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(observation) = self.input.append(value)? {
            self.observations.append(observation);
        }
        Ok(self.value())
    }

    /// Append a chronological slice through the same persistent state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        self.input.extend(values, |observation| {
            self.observations.append(observation);
            Ok(())
        })?;
        Ok(self.value())
    }

    /// Return gross profit divided by absolute gross loss.
    ///
    /// Positive-only input returns positive infinity. Empty and all-zero input
    /// return `None`; loss-only input returns zero.
    pub fn value(&self) -> Option<f64> {
        let gross_profit = self.observations.gross_gain();
        let absolute_gross_loss = -self.observations.gross_loss();
        if absolute_gross_loss > 0.0 {
            Some(gross_profit / absolute_gross_loss)
        } else if gross_profit > 0.0 {
            Some(f64::INFINITY)
        } else {
            None
        }
    }

    /// Return the current result without replaying prior observations.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving input configuration.
    pub fn reset(&mut self) {
        self.input.reset();
        self.observations.reset();
    }

    /// Return the number of usable observations processed.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable observations have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}
